File: src/engine/Media.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace SimpleEngine
{
    public abstract class Media
    {
        public abstract void load(MediaLoader ml, FileSystem fs, string path);
        protected abstract void unload();
        public void increaseUsage()
        {
            ++usage;
        }
        public void decreaseUsage()
        {
            --usage;
            if (usage == 0) unload();
        }
        private int usage = 0;
    }
}

File: src/engine/media.cc
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace SimpleEngine
{
    struct Media
    {
        void load(MediaLoader ml, FileSystem fs, std::string path);

        void unload();
        void increaseUsage()
        {
            ++usage;
        }
        void decreaseUsage()
        {
            --usage;
            if (usage == 0)
                unload();
        }

        int usage = 0;
    }
}
